### PR TITLE
Initial MSR Policy Work

### PR DESCRIPTION
**Pull request by rwdrich**
Signed-off-by: Richard Davies <richard.davies@citrix.com>

### CODE REVIEW

**Comment by rwdrich** [comment]
These should probably all be merged together:

https://github.com/lindig/xenctrl/pull/2
https://github.com/xapi-project/xcp-idl/pull/308
https://github.com/xapi-project/xenopsd/pull/706
https://github.com/xapi-project/xen-api/pull/4056

File: ocaml/xapi/cpuid_helpers.ml [hunk 1]
```diff
@@ -88,20 +74,23 @@ let features_pv       = Map_check.field Xapi_globs.cpu_info_features_pv_key feat
...
+    | `hvm | `pv_in_pvh ->
+      cpu_info_policy_hvm_key
+    | `pv ->
+      cpu_info_policy_pv_key
   in
   let vendor = List.assoc cpu_info_vendor_key cpu_info in
-  let migration = Map_check.getf features_field cpu_info in
-  let onboot = Map_check.getf ~default:migration features_field_boot cpu_info in
-  (vendor, migration, onboot)
+  let policy = List.assoc policy_key cpu_info in
```
Comment (edwintorok) [inline ocaml/xapi/cpuid_helpers.ml]: Can you use the Map_check.getf style access as above?

File: ocaml/xapi/cpuid_helpers.ml [hunk 2]
```diff
@@ -200,22 +206,17 @@ let assert_vm_is_compatible ~__context ~vm ~host ?remote () =
...
-		(string_of_features vm_cpu_features) (string_of_features host_cpu_features');
-        let vm_cpu_features' =
-          vm_cpu_features
-          |> upgrade_features ~__context ~vm host_cpu_features'
-        in
-        if not (is_subset vm_cpu_features' host_cpu_features') then begin
-          debug "VM CPU features (%s) are not compatible with host CPU features (%s)\n" (string_of_features vm_cpu_features') (string_of_features host_cpu_features');
-          debug "Host is missing these CPU features required by the VM: %s" (string_of_features (diff vm_cpu_features' host_cpu_features'));
-          fail "VM last booted on a CPU with features this host's CPU does not have."
+      if List.mem_assoc cpu_info_policy_key vm_cpu_info then begin
```
Comment (edwintorok) [inline ocaml/xapi/cpuid_helpers.ml]: I think there has to be some upgrade path  from VMs that had the old v1 or v2 style features and do not have policies yet. Is there such a function somewhere?

File: ocaml/xapi/cpuid_helpers.ml [hunk 3]
```diff
@@ -45,41 +45,27 @@ let zero_extend arr len =
...
-  is_equal (intersect left right) left
-
-(** is_strict_subset left right returns true if left is a strict subset of right
-    (left is a subset of right, but left and right are not equal) *)
-let is_strict_subset left right =
-  (is_subset left right) && (not (is_equal left right))
+let intersect ~__context left right =
+  let open Xapi_xenops_queue in
+  let dbg = Context.string_of_task __context in
+  let module Client = (val make_client (default_xenopsd ()): XENOPS) in
```
Comment (lippirk) [inline ocaml/xapi/cpuid_helpers.ml]: you could create this module once in the global scope of this file? or does making the client have a side effect?

File: ocaml/xapi/cpuid_helpers.ml [hunk 4]
```diff
@@ -45,41 +45,27 @@ let zero_extend arr len =
...
+  Client.HOST.policy_calc_compatible dbg left right
+
+let is_equal ~__context left right =
+  let open Xapi_xenops_queue in
+  let dbg = Context.string_of_task __context in
+  let module Client = (val make_client (default_xenopsd ()): XENOPS) in
+  Client.HOST.policy_is_equal dbg left right
```
Comment (rwdrich) [inline ocaml/xapi/cpuid_helpers.ml]: policy_is_equal won't exist

**Pull request closed without merging**